shared/src/main/scala/org/atnos/eff/ListEffect.scala

```diff
@@ -41,69 +41,28 @@ trait ListCreation {
 object ListCreation extends ListCreation
 
 trait ListInterpretation {
+  
   /** run an effect stack starting with a list effect */
-  def runList[R, U, A](effects: Eff[R, A])(implicit m: Member.Aux[List, R, U]): Eff[U, List[A]] = {
-    val loop = new Loop[List, R, A, Eff[U, List[A]], Eff[U, Unit]] {
-      type S = (List[Eff[R, A]], ListBuffer[A])
-      val init = (List[Eff[R, A]](), new ListBuffer[A])
-
-      def onPure(a: A, s: S): (Eff[R, A], S) Either Eff[U, List[A]] =
-        s match {
-          case (head :: tail, result) => Left((head, (tail, result :+ a)))
-          case (List(), result)       => Right(EffMonad[U].pure((result :+ a).toList))
+  def runList[R, U, A](effect: Eff[R, A])(implicit m: Member.Aux[List, R, U]): Eff[U, List[A]] =
+    runInterpreter(effect)(new Interpreter[List, U, A, List[A]] {
+      def onPure(a: A): Eff[U, List[A]] =
+        Eff.pure(List(a))
+
+      def onEffect[X](xs: List[X], continuation: Continuation[U, X, List[A]]): Eff[U, List[A]] =
```

Comment:
That's a good idea. We could even have a `TraverseEffect` to generalize the `ListEffect`.